### PR TITLE
rwlock: Move over to condvar implementation.

### PR DESCRIPTION
**Pull request by plutooo**
This one is superior to the previous because it is write-preferring.

### CODE REVIEW

File: nx/include/switch/kernel/rwlock.h [hunk 1]
```diff
@@ -6,12 +6,15 @@
...
 /// Read/write lock structure.
 typedef struct {
-    RMutex r;
-    RMutex g;
-    u64 b;
+    Mutex mutex;
+    CondVar condvar_readers;
+    CondVar condvar_writer;
+    u64  readers;
+    bool writer;
```
Comment (fincs) [inline nx/include/switch/kernel/rwlock.h]: Save space and avoid alignment padding by using bitfields:
```c
    u32 readers : 31;
    bool writer : 1;
```

File: nx/source/kernel/rwlock.c [hunk 2]
```diff
@@ -3,33 +3,57 @@
...
+    mutexUnlock(&r->mutex);
 }
 
 void rwlockReadUnlock(RwLock* r) {
-    rmutexLock(&r->r);
+    mutexLock(&r->mutex);
 
-    if (r->b-- == 1)
-        rmutexUnlock(&r->g);
+    if (r->readers-- == 1) {
```
Comment (fincs) [inline nx/source/kernel/rwlock.c]: `if (!--r->readers)`

File: nx/source/kernel/rwlock.c [hunk 3]
```diff
@@ -3,33 +3,57 @@
...
-    rmutexLock(&r->g);
+    mutexLock(&r->mutex);
+
+    while (r->writer) {
+        condvarWait(&r->condvar_writer, &r->mutex);
+    }
+
+    r->writer = true;
+
+    while (r->readers > 0) {
```
Comment (fincs) [inline nx/source/kernel/rwlock.c]: `while (r->readers)`

**Comment by fincs** [comment]
Pinging @plutooo.

File: nx/include/switch/kernel/rwlock.h (hunk 1 above)
Comment (plutooo) [reply nx/include/switch/kernel/rwlock.h]: fixed

File: nx/source/kernel/rwlock.c (hunk 2 above)
Comment (plutooo) [reply nx/source/kernel/rwlock.c]: fixed-ish, prefer not using boolean operators (!) on integer types

File: nx/source/kernel/rwlock.c (hunk 3 above)
Comment (plutooo) [reply nx/source/kernel/rwlock.c]: prefer not using boolean operators (!) on integer types

File: nx/source/kernel/rwlock.c (hunk 2 above)
Comment (fincs) [reply nx/source/kernel/rwlock.c]: That's fine. However, those parentheses are superflous and you can remove them: `if (--r->readers == 0)` is also correct.

**Pull request merged**